src/main/java/org/spongepowered/common/data/processor/value/entity/BaseVehicleValueProcessor.java

```diff
@@ -0,0 +1,133 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.common.data.processor.value.entity;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.base.Function;
+import com.google.common.base.Optional;
+import net.minecraft.entity.EntityLivingBase;
+import org.spongepowered.api.data.DataTransactionBuilder;
+import org.spongepowered.api.data.DataTransactionResult;
+import org.spongepowered.api.data.key.Key;
+import org.spongepowered.api.data.key.Keys;
+import org.spongepowered.api.data.value.BaseValue;
+import org.spongepowered.api.data.value.ValueContainer;
+import org.spongepowered.api.data.value.immutable.ImmutableValue;
+import org.spongepowered.api.data.value.mutable.Value;
+import org.spongepowered.api.entity.Entity;
+import org.spongepowered.common.data.ValueProcessor;
+import org.spongepowered.common.data.value.immutable.ImmutableSpongeValue;
+import org.spongepowered.common.data.value.mutable.SpongeValue;
+
+public class BaseVehicleValueProcessor implements ValueProcessor<Entity, Value<Entity>> {
+
+    @Override
+    public Key<? extends BaseValue<Entity>> getKey() {
+        return Keys.BASE_VEHICLE;
+    }
+
+    @Override
+    public Optional<Entity> getValueFromContainer(ValueContainer<?> container) {
+        if (container instanceof net.minecraft.entity.Entity) {
+            net.minecraft.entity.Entity currentEntity = (net.minecraft.entity.Entity) container;
+            while (currentEntity != null && currentEntity.isRiding())
+            {
+                currentEntity = currentEntity.ridingEntity;
+            }
+            final Entity baseVehicle = (Entity) currentEntity;
+            return Optional.of(baseVehicle);
+        }
+        return Optional.absent();
+    }
+
+    @Override
+    public Optional<Value<Entity>> getApiValueFromContainer(ValueContainer<?> container) {
+        if (container instanceof net.minecraft.entity.Entity) {
+            net.minecraft.entity.Entity currentEntity = (net.minecraft.entity.Entity) container;
+            while (currentEntity != null && currentEntity.isRiding())
+            {
+                currentEntity = currentEntity.ridingEntity;
+            }
+            final Entity baseVehicle = (Entity) currentEntity;
```

Comment:
You should double check that the currentEntity isn't the same as the container, because otherwise, you're always returning a "base vehicle" of the entity itself, which in most cases, isn't true.